### ISSUE TITLE
Build only for merge requests, not for regular pushes

### ISSUE THREAD

**Issue by Yogu** [jenkinsci/gitlab-plugin#49]
We have two jobs on our servers:

* a normal job that builds `master`, so we always see if the main branch passes the build and have a nice history of code metrics
* the merge-request job, configured as instructed in this README, to get build information in the merge requests

Currently, the merge-request job also builds the branches on normal pushes, so on every push to `master`, two builds are scheduled. We'd like the merge-request job to only build when pushing to a branch with an open merge request, i.e. iff the build comment will be of the form "GitLab ​Merge ​​Request ​​#xx ​​: source-branch ​​=> ​​master ".

Is this possible? If I disable "Build on Push Events", subsequent pushes to open merge requests are not built. The "Filter branches" box also looks promising, we could simply exclude master, but the list is empty for me, even after a few builds.

**Comment by kasper-f** [comment]
I haven't been able to reproduce the issue with duplicate builds on push. 

I have test with
* uncheck the config box "Build on Merge Request Events" (all other checked) and the push to a branch with an open merge request. It did not trigger a new build.
*  uncheck the config box "Build on Push Events" (all other checked)  and push to master branch. It did not trigger a new build.

In my config I have selected to only build the "master" branch in the "Filter branches" option. I don't know why you don't get a list of branches in the filter branch config option. Have you tried to look at Jenkins log file to see if it give you any hints regarding the missing branches?

**Comment by MartinNowak** [comment]
It works as supposed out of the box, just disable the build on push.
Your problem might be a missing global gitlab configuration b/c the webhook needs to query gitlab's api in order to get the list of open merge requests. Just check the jenkins 'system log' and/or add the gitlab api token in 'configure system'.

**Comment by WonderCsabo** [comment]
I also have this problem. I would not like to disable build on push because i want to build all branches, but if a branch is on a merge request, do not duplicate the build (just build the MR event instead).

**Comment by sbamani** [comment]
I have a similar problem - even when I uncheck the "Build when changes are pushed to gitlab" it is still building on every push... I tried checking it and the uncheck the "Build on Merge Request Events" and the "Build on Push Events" But it still trigger a build.

**Comment by coder-hugo** [comment]
Duplicate of #210

**Comment by starthal** [comment]
Is the behavior described here still expected/intended? 

When MR trigger and push trigger are both enabled, a push to an open MR should trigger two builds (one merged, one not merged)?

This "duplication" of the builds doesn't bother me, but they both seem to update the MR pipeline status in GitLab, so there is a race condition when the builds are run in parallel. If one fails and the other succeeds, the result is confusing.

**Comment by omehegan** [comment]
@starthal yes I think it is expected. IIRC they are technically two different commits so the plugin doesn't have any way to discern that they are or could be the same effective state of the repo.

**Comment by starthal** [comment]
@omehegan Thanks. Is there a way to prevent the non-merged build from updating the MR status in GitLab? Or to give different "build names" to the merged and non-merged builds so GitLab treats them as separate "pipelines"?

By "build name" I mean this field:

![2018-01-15-084746_580x164_scrot](https://user-images.githubusercontent.com/3247950/34945278-ccc6a054-f9d0-11e7-82d7-d0c6a902af94.png)

**Comment by omehegan** [comment]
@starthal if it was a Pipeline job, you could use some logic to look at the gitlabActionType env var and only update status for one or the other type.
